File: erpnext/selling/report/sales_person_commission_summary/sales_person_commission_summary.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt


import frappe
from frappe import _, msgprint, qb
from frappe.query_builder import Criterion


def execute(filters=None):
	if not filters:
		filters = {}

	columns = get_columns(filters)
	entries = get_entries(filters)
	data = []

	for d in entries:
		data.append(
			[
				d.name,
				d.customer,
				d.territory,
				d.posting_date,
				d.base_net_amount,
				d.sales_person,
				d.allocated_percentage,
				d.commission_rate,
				d.allocated_amount,
				d.incentives,
			]
		)

	if data:
		total_row = [""] * len(data[0])
		data.append(total_row)

	return columns, data


def get_columns(filters):
	if not filters.get("doc_type"):
		msgprint(_("Please select the document type first"), raise_exception=1)

	columns = [
		{
			"label": _(filters["doc_type"]),
			"options": filters["doc_type"],
			"fieldname": filters["doc_type"],
			"fieldtype": "Link",
			"width": 140,
		},
		{
			"label": _("Customer"),
			"options": "Customer",
			"fieldname": "customer",
			"fieldtype": "Link",
			"width": 140,
		},
		{
			"label": _("Territory"),
			"options": "Territory",
			"fieldname": "territory",
			"fieldtype": "Link",
			"width": 100,
		},
		{"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 100},
		{"label": _("Amount"), "fieldname": "amount", "fieldtype": "Currency", "width": 120},
		{
			"label": _("Sales Person"),
			"options": "Sales Person",
			"fieldname": "sales_person",
			"fieldtype": "Link",
			"width": 140,
		},
		{
			"label": _("Contribution %"),
			"fieldname": "contribution_percentage",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"label": _("Commission Rate %"),
			"fieldname": "commission_rate",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("Contribution Amount"),
			"fieldname": "contribution_amount",
			"fieldtype": "Currency",
			"width": 120,
		},
		{"label": _("Incentives"), "fieldname": "incentives", "fieldtype": "Currency", "width": 120},
	]

	return columns


def get_entries(filters):
	dt = qb.DocType(filters["doc_type"])
	st = qb.DocType("Sales Team")
	date_field = dt["transaction_date"] if filters["doc_type"] == "Sales Order" else dt["posting_date"]

	conditions = get_conditions(dt, st, filters, date_field)
	entries = (
		qb.from_(dt)
		.join(st)
		.on(st.parent.eq(dt.name) & st.parenttype.eq(filters["doc_type"]))
		.select(
			dt.name,
			dt.customer,
			dt.territory,
			date_field.as_("posting_date"),
			dt.base_net_total.as_("base_net_amount"),
			st.commission_rate,
			st.sales_person,
			st.allocated_percentage,
			st.allocated_amount,
			st.incentives,
		)
		.where(Criterion.all(conditions))
		.orderby(dt.name, st.sales_person)
		.run(as_dict=True)
	)

	return entries


def get_conditions(dt, st, filters, date_field):
	conditions = []

	conditions.append(dt.docstatus.eq(1))
	from_dt = filters.get("from_date")
	to_dt = filters.get("to_date")
	if from_dt and to_dt:
		conditions.append(date_field.between(from_dt, to_dt))
	elif from_dt and not to_dt:
		conditions.append(date_field.gte(from_dt))
	elif not from_dt and to_dt:
		conditions.append(date_field.lte(to_dt))

	for field in ["company", "customer", "territory"]:
		if filters.get(field):
			conditions.append(dt[field].eq(filters.get(field)))

	if filters.get("sales_person"):
		conditions.append(st["sales_person"].eq(filters.get("sales_person")))

	return conditions

File: erpnext/selling/report/sales_person_commission_summary/sales_person_commission_summary.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

const group_by_options_spcs = [
	"Ungrouped", "Group by Sales Person", "Group by Territory", "Group by Sales Commission Category",
	"Group by Invoice", "Group by Customer"
]

frappe.query_reports["Sales Person Commission Summary"] = {
	"filters": [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "date_type",
			label: __("Which Date"),
			fieldtype: "Select",
			options: [
				'Invoice Date',
				'Clearing Date'
			],
			default: "Invoice Date",
			reqd: 1,
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.defaults.get_user_default("year_start_date"),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.defaults.get_user_default("year_end_date"),
			reqd: 1,
		},
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "Link",
			options: "Sales Person",
			bold: 1,
		},
		{
			fieldname: "territory",
			label: __("Territory"),
			fieldtype: "Link",
			options: "Territory",
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "name",
			label: __("Sales Invoice"),
			fieldtype: "Link",
			options: "Sales Invoice",
			filters: {
				"docstatus": 1,
			}
		},
		{
			fieldname: "sales_commission_category",
			label: __("Sales Commission Category"),
			fieldtype: "Link",
			options: "Sales Commission Category"
		},
		{
			fieldname: "transaction_type",
			label: __("Transaction Type"),
			fieldtype: "Link",
			options: "Transaction Type"
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Cost Center', txt, {
					company: frappe.query_report.get_filter_value("company")
				});
			}
		},
		{
			fieldname: "group_by_1",
			label: __("Group By Level 1"),
			fieldtype: "Select",
			options: group_by_options_spcs,
			default: "Ungrouped"
		},
		{
			fieldname: "group_by_2",
			label: __("Group By Level 2"),
			fieldtype: "Select",
			options: group_by_options_spcs,
			default: "Ungrouped"
		},
		{
			fieldname: "totals_only",
			label: __("Group Totals Only"),
			fieldtype: "Check",
			default: 0
		},
		{
			fieldname: "exclude_unpaid_invoices",
			label: __("Exclude Unpaid Invoices"),
			fieldtype: "Check",
		},
		{
			fieldname: "exclude_zero_commission",
			label: __("Exclude Zero Commission Rate"),
			fieldtype: "Check",
		},
		{
			fieldname: "show_deduction_details",
			label: __("Show Deduction Details"),
			fieldtype: "Check",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		var style = {};

		if (["net_contribution_amount", "commission_amount"].includes(column.fieldname)) {
			style['font-weight'] = 'bold';
		}

		if (flt(value) && (column.fieldname == "total_deductions" || column.is_adjustment)) {
			style['color'] = 'orange';
		}

		if (flt(value) && column.fieldname == "outstanding_amount") {
			style['color'] = 'red';
		}

		if (flt(value) && column.fieldname == "paid_amount") {
			style['color'] = 'green';
		}

		if (flt(value) && column.fieldname == "return_amount") {
			style['color'] = 'orange';
		}

		if (value && flt(value) <= 0 && column.fieldname == "age") {
			style['color'] = 'blue';
		}

		if (flt(value) && column.fieldname == "late_payment_deduction_percent") {
			style['color'] = 'red';
		}

		if (data && flt(data.base_net_amount) < 0) {
			if (['name', 'base_net_amount', 'contribution_amount', 'net_contribution_amount', 'commission_amount',
					'base_grand_total'].includes(column.fieldname)) {
				style['color'] = 'red';
			}
		}

		return default_formatter(value, row, column, data, {css: style});
	},
	"initial_depth": 1,
}
